src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1291,7 +1272,67 @@ public Boolean visitUndetVar(UndetVar t, Type s) {
             public Boolean visitErrorType(ErrorType t, Type s) {
                 return true;
             }
-        };
+        }
+
+        public NullabilityComparator nullabilityComparator = new NullabilityComparator();
+        public class NullabilityComparator extends TypeRelation {
+            BiFunction<Type, Type, Boolean> differentNullability;
+            Warner warner;
+
+            NullabilityComparator reset(BiFunction<Type, Type, Boolean> differentNullability) {
+                this.differentNullability = differentNullability;
+                if (this.warner == null || this.warner == noWarnings) {
```

Comment:
I'm not sure I follow what's happening here. We have a warner stack in Types already. But this class seems to somehow override whatever warner is in the stack (if a client called setWarner on the nullability comparator). Shouldn't we try to reuse the warnStack as a way to set warners, and then tweak nullability comparator to use whatever warner is on the stack? E.g. I'm suspicious about the set/clearWarner methods, as they seem to duplicate functionality we already have in other forms.

Reply:
I did this because Check::checkExtends invokes Types::isSubtype, in a previous iteration I think you suggested not to add a Warner argument to isSubtypes, so this is an effort to circumvent that. I will give it another try